src/main/java/com/commercetools/pspadapter/notification/processor/impl/PaymentSaleNotificationProcessorBase.java

```diff
@@ -35,8 +47,37 @@ public boolean canProcess(@Nonnull Event event) {
                 .equalsIgnoreCase(event.getEventType());
     }
 
-    protected AddTransaction createAddTransactionAction(@Nonnull Event event,
-                                                        @Nonnull TransactionType transactionType) {
+    protected List<? extends UpdateAction<Payment>> createUpdatePaymentActions(@Nonnull Payment ctpPayment, @Nonnull Event event) {
+        String resourceId = getResourceId(event);
+        return findTransactionByInteractionId(ctpPayment.getTransactions(), resourceId)
+                .map(txn -> {
+                    if (!txn.getType().equals(getExpectedTransactionType())) {
+                        logger.error("Found txn paymentId=[{}] with corresponding resourceId={},"
+                                        + " but transactionType=[{}] is not expectedTransactionType=[{}]."
+                                        + " Will create new transaction for the eventId=[{}]",
+                                ctpPayment.getId(), resourceId, getExpectedTransactionType(), event.getId());
+                        return createAddTransactionActionList(event, getExpectedTransactionType());
+                    } else if (isTxnAlreadyUpdated(txn)) {
+                        // can't do Collections.emptyList() here because map() and generics
+                        // together generates unexpected return results
+                        return new ArrayList<UpdateAction<Payment>>(0);
+                    } else {
+                        return createChangeTransactionStateActionList(txn);
+                    }
+                })
+                .orElseGet(() -> createAddTransactionActionList(event, getExpectedTransactionType()));
+    }
+
+    abstract protected TransactionType getExpectedTransactionType();
```

Comment:
returns `@Nonnul`, same for `TransactionState getExpectedTransactionState();` below

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/69/commits/f709e3de8f1e8edbc9730da0820ff4d24e9b73cd